gpflow/training/monitor.py

```diff
@@ -621,6 +640,90 @@ def run(self, context: MonitorContext, *args, **kwargs) -> None:
                          global_step=context.global_step_tensor)
 
 
+class LogdirWriter(tf.summary.FileWriter):
+    """
+    This is a wrapper around the TensorFlow summary.EventWriter that provides a workaround for
+    a bug currently present in this module. The EventWriter can only open the file in exclusive
+    mode, however when multiple instances of the writer attempt to access the same file no error
+    is raised.
+
+    This class prevents user from opening multiple writers with the same location (event file
+    directory and file name suffix). It keeps a global set of used locations adding a new location
+    there when a writer is created or reopened. It removes the location from the global set when
+    the writer is closed or garbage collected.
+
+    Once the bug in TensorFlow is fixed this class can be removed or reduced to trivial:
+    class LogdirWriter(tf.summary.FileWriter):
+        pass
+    """
+
+    _used_locations = set()     # type: Set[Tuple[str, Optional[str]]
+
+    def __init__(self, logdir: str, graph: Optional[tf.Graph]=None, max_queue: int=10,
+                 flush_secs: float=120, filename_suffix: Optional[str]=None):
+        """
+        A thin wrapper around the summary.FileWriter __init__. It remembers the location (a tuple
+        of event file directory and file name suffix) and attempts to lock it. If the location
+        is already locked by another writer an error will be raised.
+
+        :param logdir: Directory where event file will be written.
+        :param graph: A `Graph` object
+        :param max_queue: Size of the queue for pending events and summaries.
+        :param flush_secs: How often, in seconds, to flush the added summaries and events to disk.
+        :param filename_suffix: Optional suffix of the event file's name.
+        """
+
+        self._location = (str(PurePath(logdir)), filename_suffix)
+        self._is_locked = False
+        self.__lock_location()
+        super().__init__(logdir, graph, max_queue, flush_secs, filename_suffix=filename_suffix)
+
+    def __del__(self):
+        self.__release_location()
+        if hasattr(super(), '__del__'):
```

Comment:
All objects have __del__ method. It is not necessary to check it.

Reply:
I don't think it's true. Try this
`hasattr(object(), '__del__')`

Reply:
Yep, remove `hasattr(object(), '__del__')` with super().__del__()

Reply:
@ahsimb Now, I see why `_is_locked` is needed, the name is a bit confusing. I would rename it to `_active` - like you are activating this writer, when you close it you deactivate it and __del__ will do nothing, but in the same time another writer could be created and place same logdir in the set.

Reply:
`del` is not a method of an object. It's a Python statement that deletes (garbage collects) an object. A class may implement a destructor `__del__`. The destructor, if defined, will be called when the object is removed either by `del` or GC. The parent class could have defined its own destructor which should also be called to clear the parent's resources. I believe the way I've done this is correct.

Reply:
I know that `__del__` is not part of the object. What I meant is that `__del__` is not overwritten by FileWriter and therefore you don't need to call `hasattr(object(), '__del__')` and `super().__de__()` respectively.